data/binding/binding.go

```diff
@@ -0,0 +1,58 @@
+//go:generate go run gen.go
+
+package binding
+
+// DataItem is the base interface for all bindable data items.
+type DataItem interface {
+	AddListener(DataItemListener)
+	RemoveListener(DataItemListener)
+}
+
+// DataItemListener is any object that can register for changes in a bindable DataItem.
+// See NewDataItemListener to define a new listener using just an inline function.
+type DataItemListener interface {
+	DataChanged(DataItem)
+}
+
+// NewDataItemListener is a helper function that creates a new listener type from a simple callback function.
+func NewDataItemListener(fn func(DataItem)) DataItemListener {
+	return &listener{fn}
+}
+
+type listener struct {
+	callback func(DataItem)
+}
+
+func (l *listener) DataChanged(i DataItem) {
+	l.callback(i)
+}
+
+type base struct {
+	listeners []DataItemListener
```

Comment:
Reads and writes to this slice should be locked

Reply:
yup thanks